Align JwtClient authentication with latest OIDC spec updates

thomasdarimont opened the pull request:
This PR introduces new "lenient-audience-validation" config property with default "false" to JWTClientAuthenticator and JWTClientSecretAuthenticator. The new property allows users to require that only the issuer URL is allowed in the audience `aud` claim of a JWT client assertion used for `private_key_jwt` and `client_secret_jwt` client authentication methods. By default we enforce the new behavior, server admins can revert to the old more lenient `aud` claim checks via the new client authenticator configuration properties and a spi configuration property.

To use lenient audience validation for the server by default, users can set the following properties:
```
--spi-client-authenticator-client-jwt-lenient-audience-validation=true
--spi-client-authenticator-client-secret-jwt-lenient-audience-validation=true
```

This aligns with changes in the OIDC core specification https://openid.net/specs/openid-connect-core-1_0-36.html#rfc.section.9 for the aud validation for `private_key_jwt` and `client_secret_jwt` client authentication methods.

Fixes #38751

Config Option for Signed JWT Authenticator
<img width="420" alt="image" src="https://github.com/user-attachments/assets/b9ffbf3f-9a8c-42f9-8f71-03a264506c49" />

Config option for Signed JWT with Client Secret Authenticator
<img width="419" alt="image" src="https://github.com/user-attachments/assets/3de8a825-7a5c-4dfc-84b8-31df97db1087" />


<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
Align JwtClient authentication with latest OIDC spec updates

Introduce new "issuer-only-audience" config property with default "false"
to JWTClientAuthenticator and JWTClientSecretAuthenticator.
The new property allows users to require that only the issuer URL is allowed in the
audience `aud` claim of a JWT client assertion used for `private_key_jwt` and `client_secret_jwt` client authentication methods.

This aligns with changes in the OIDC core specification https://openid.net/specs/openid-connect-core-1_0-36.html#rfc.section.9 for the aud validation for
`private_key_jwt` and `client_secret_jwt` client authentication methods.

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

rmartinc commented on services/src/main/java/org/keycloak/authentication/authenticators/client/AbstractJwtClientAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
+        return List.copyOf(configProperties);
```
Just a nitpick, but here usually we create the properties on the fly using the `ProviderConfigurationBuilder`, check for example [here](https://github.com/keycloak/keycloak/blob/26.1.4/services/src/main/java/org/keycloak/authentication/authenticators/resetcred/ResetCredentialEmail.java#L200).

thomasdarimont replied:
Thanks I'll inline the configuration list then.

rmartinc reviewed (changes requested):
Thanks for the PR @thomasdarimont! I really doubt about the default. If the spec changed because of security, keycloak should be secure by default. So having the default to `false` is not a solution. I see two options:

1. Change the property to be the other way around. The default `false` will use new spec behavior, the `true` will be the legacy behavior (changing descriptions and labels). This breaks compatibility, and the admins should change the value if necessary after upgrade.
2. Use current approach but at least for the new realms (or when the `clients` flow needs to be created in `DefaultAuthenticationFlows`), create the flow with the config to `true`. This way we are secure by default, but migrated installations continue having the legacy behavior.

As you commented, we need upgrading notes for this too. Depending if we choose 1 or 2, it would be a breaking or notable change.

@mposolda Please, take a look to this.

thomasdarimont commented:
Thanks @rmartinc for your thoughts!

Regarding your suggested options:

> 1. Change the property to be the other way around. The default false will use new spec behavior, the true will be the legacy behavior (changing descriptions and labels). This breaks compatibility, and the admins should change the value if necessary after upgrade.

I agree with you that we should switch the default, however I was just thinking about setting `ISSUER_ONLY_AUDIENCE_PROPERTY_DEFAULT=true` in `AbstractJwtClientAuthenticator` and effectivly use `issuer-only-audience=true` by default.
User who want to opt-out could then disable the setting in their configurations. However some clients might not be able to do that, therefore it might be better to have a global config property for setting.


> 2. Use current approach but at least for the new realms (or when the clients flow needs to be created in DefaultAuthenticationFlows), create the flow with the config to true. This way we are secure by default, but migrated installations continue having the legacy behavior.
If we just change the default to `issuer-only-audience=true` in `AbstractJwtClientAuthenticator`, then the client authentication flow defined by org.keycloak.models.utils.DefaultAuthenticationFlows#clientAuthFlow would be secure out of the box, without the explicit configuration.

However, some users might not be able to reconfigure the auth flow in the realm or might not be able to change client implementations.
Therefore I suggest to allow users to configure the default behavior with a command line option like the following:
```
// This allows to override the server-wide default for JWT client authentication, via command line options.
--spi-client-authenticator-client-jwt-issuer-only-audience=false
--spi-client-authenticator-client-secret-jwt-issuer-only-audience=false
```
With this users can use the old behavior for the server without any additional configuration.

If nothing is configured, we use the secure default which restricts the audience to be the issuer only audience.

thomasdarimont committed:
Revise AbstractJwtClientAuthenticator

- use `issuer-only-audience=true` by default
- Allow users to customize the default via command line options, e.g. to use relaxed audience validation:
```
// --spi-client-authenticator-client-jwt-issuer-only-audience=false
// --spi-client-authenticator-client-secret-jwt-issuer-only-audience=false
```
With this users can gradually adopt the new behavior.

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
The tests are probably failing now, because the issuer-only-audience=true behavior is now effective.
I'd suggest to configure the respective tests to use the following properties for now.
```
--spi-client-authenticator-client-jwt-issuer-only-audience=false
--spi-client-authenticator-client-secret-jwt-issuer-only-audience=false
```

How can I do that?

The affected tests are:
- ClientAuthSignedJWTTest
- ClientAuthSecretSignedJWTTest

rmartinc commented:
@thomasdarimont So I understand that you are going directly with solution 1 (I'm O with it). I prefer to use default of `false` always for any option, but it's just a personal opinion. I think that it's clear that false is the default, the same that no defining any configuration.

For the test I think that the easiest path is adding code to add the property in the authenticator for those tests at authenticator level. If you want to test also the startup option you need to use the quarkus tests, but you can skip this part if you want.

thomasdarimont commented:
Thanks again @rmartinc! I'll invert the logic of the switch and change the default from `issuer-only-audience=true` to `lenient-audience-validation: false` and adapt the code / tests to the new semantics.

thomasdarimont committed:
Change default from issuer-only-audience=true to lenient-audience-validation=false

- Adapt code and tests
- Add release note

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

rmartinc commented on docs/documentation/release_notes/topics/26_2_0.adoc:
```diff
@@ -96,6 +96,21 @@ For more information, check the link:https://www.keycloak.org/server/management-
 Introduced the ability to dynamically select authentication flows based on conditions such as requested scopes, ACR (Authentication Context Class Reference) and others.
 This can be achieved using link:{adminguide_link}#_client_policies[Client Policies] by combining the new `AuthenticationFlowSelectorExecutor` with conditions like the new `ACRCondition`. For more details, see the link:{adminguide_link}#_client-policy-auth-flow[{adminguide_name}].
 
+= JWT Client authentication aligned with the latest OIDC specification
+
+The latest version of the link:https://openid.net/specs/openid-connect-core-1_0-36.html#rfc.section.9[OpenID Connect core specification] changed the rules for
+audience validation in JWT client assertions for the Client Authentication methods `private_key_jwt` and `client_secret_jwt`.
+
+Previously, the `aud` claim of a JWT client assertion was loosely defined as `The Audience SHOULD be the URL of the Authorization Server's Token Endpoint`, which did not exclude the usage of other URLs.
+
+The revised OIDC Core specification uses a stricter audience check: "The Audience value MUST be the OP's Issuer Identifier passed as a string, and not a single-element array.".
+
+We adapted the implementations of `JWTClientAuthenticator` and `JWTClientSecretAuthenticator` to follow the new audience check rules.
+By default, we now require the issuer URL to be the sole value of the audience.
+
+This strict audience check can be reverted to the previous more lenient check with a new 'Lenient Audience Validation'
+configuration in the respective client authentication flows. Additionally, server administrators can use the SPI configuration options `--spi-client-authenticator-client-jwt-lenient-audience-validation=true`and `--spi-client-authenticator-client-secret-jwt-lenient-audience-validation=true` to configure the old audience validation behavior.
+
```
@thomasdarimont We are using the upgrading notes for [breaking changes](https://github.com/keycloak/keycloak/blob/main/docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc#breaking-changes). So please move this part to that section. If you want you can maintain here a little note but you can just move everything there.

thomasdarimont replied:
Alright, I'll move the paragraph over and add a link to the upgrading guide with more details.

rmartinc reviewed (changes requested):
Thanks @thomasdarimont! It's OK to me at code level. Just a nitpick in the tests. But, please, move the release note to the upgrading section instead. We are using that part for breaking changes like this one.

@mposolda Once the doc is moved to the upgrading guide is OK for me. Probably I cannot check this in the rest of the day.

thomasdarimont committed:
Adapt code and documentation according to review

- Add section to changes document
- Add brief section to release document
- Revise method name in AbstractClientAuthSignedJWTTest

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
Just updated the PR.

Thanks for your suggestions @mposolda. If possible I'd like to keep the SPI configuration in as well, as there might be cases where the Keycloak server admins cannot change the realm configuration, but need to to an upgrade for security reasons.

In order to avoid breaking existing clients, server admins can then use the cli config to configure the server-wide default for the aud validation. Note Keycloak administrators with realm access can configure the client authentication flow as needed per realm.

thomasdarimont committed:
Adapt OIDCJwksClientRegistrationTest to use the realm issuer as audience for private_key_jwt client auth

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont committed:
Adapt AbstractOAuth2IdentityProvider and OIDCIdentityProvider to stricter audience validation for private_key_jwt client auth method.

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
While fixing the failing tests I noticed that there is another subtle issue here:
The test `KcOidcBrokerPrivateKeyJwtUnsupportedKeyTest` failed because another Keycloak realm was used as an IdentityProvider and ueses private_key_jwt for client authentication.

However, the current implementation of org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider#generateToken  used the token endpoint as the audience for the generated JWT client assertion.
I extracted the JWT client assertion audience creation into a new method and added another implementation here: org.keycloak.broker.oidc.OIDCIdentityProvider#createJwtClientAssertionAudience. In the new method I explicitly use the issuer if this is an OIDCIdentityProvider provider.

Clients that still want to keep using the token endpoint, could explicitly configure the clientAssertionAudience to use a custom value.

See:
https://github.com/keycloak/keycloak/commit/44d19a986c03937da5bf51031724ca0dc3546a63

mposolda commented:
> Just updated the PR.
> 
> Thanks for your suggestions @mposolda. If possible I'd like to keep the SPI configuration in as well, as there might be cases where the Keycloak server admins cannot change the realm configuration, but need to to an upgrade for security reasons.
> 
> In order to avoid breaking existing clients, server admins can then use the cli config to configure the server-wide default for the aud validation. Note Keycloak administrators with realm access can configure the client authentication flow as needed per realm.

Sounds good. Thanks for the clarification. Since both you and @rmartinc think that config in your PR is ok, I agree with keeping as is.

> However, the current implementation of org.keycloak.broker.oidc.AbstractOAuth2IdentityProvider#generateToken used the token > endpoint instead of the issuer as the audience for the generated JWT client assertion.
> I extracted the JWT client assertion audience creation into a new method and added another implementation here: org.keycloak.broker.oidc.OIDCIdentityProvider#createJwtClientAssertionAudience. In the new method I explicitly use the issuer if > this is an OIDCIdentityProvider provider.

> Clients that still want to keep using the token endpoint, could explicitly configure the clientAssertionAudience to use a custom value.

> See:
> https://github.com/keycloak/keycloak/commit/44d19a986c03937da5bf51031724ca0dc3546a63

This is good catch for identity brokering!

IMO this will be also good to document in the upgrading guide, that OIDC Identity providers use `issuer` by default instead of token endpoint unless it is overriden by `clientAssertionAudience` . As 3rd party IDP may still expect token-endpoint being used as an `audience` and hence client authentication may suddently stop working. Will be good to also update the tooltip of the related option in admin console and mention that "issuer" is the default for OIDC Identity providers instead of "token URL" : https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties#L850 .

Question: How about change this behaviour directly in `AbstractOAuth2IdentityProvider.createJwtClientAssertionAudience()` and use the `issuer` here as well instead, so there won't be a need to override `createJwtClientAssertionAudience()` for `OIDCIdentityProvider`? This means there would be the same behaviour for "OAuth2" and "OIDC" identity providers. Or any reason why OAuth2 providers should be different and still use token endpoint?

thomasdarimont committed:
Document changed JWT Client Assertion for OIDC Identity Provider.

Add a note to changes about the new audience in JWT Client Assertion for OIDC Identity Providers.

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

rmartinc commented:
@mposolda I think the reason is that the `issuer` configuration is only in the OIDC config and not in the oauth2 config. So I think that is OK. In the OIDC configuration there is also a field to change the audience if they need to modify it.

@thomasdarimont One last comment about this last change. Probably we need to change the documentation [here](https://github.com/keycloak/keycloak/blob/main/docs/documentation/server_admin/topics/identity-broker/oidc.adoc?plain=1#L50) and the tool-tip [here] (https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties#L850). Saying that `the default value is the OpenIDC provider's Issuer as per specification` or similar.

I see you have added also this chnage in the breaking changes. So that's all.

thomasdarimont commented:
I just added an additional note to the changes document about the changed audience value for OIDC Identity Providers that use JWT based client authentication -> using issuer in the audience instead of the token url.

Users can use the `Client Assertion audience` property in their OIDC identity provider configuration, to revert to the old behavior.
![image](https://github.com/user-attachments/assets/c502fa86-e5a3-4071-8747-609fa715536b)

thomasdarimont committed:
Adapt translations and help texts for changes in JWT Client assertion audience.

Fixes #38751

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
@rmartinc just pushed the translation and doc updates you mentioned.

mposolda reviewed (approved):
@thomasdarimont Latest updates in the documentation and tooltip looks OK to me. Thanks! Approving

stianst commented:
It's a shame that this has even been considered as a CVE; as it really is not. It's rather a hardening thing. There's nothing novel or unexpected with the fact that if you include multiple audiences in a token it can be abused if one of the audiences are compromised.

Another complicating fact is the change of the URL from the token endpoint (which never made any sense though) to the issuer URL.

Finally, the changes in the spec are still in a draft state; and the current [non-draft](https://openid.net/specs/openid-connect-core-1_0.html) does not include this change, which means anyone referring to the non-draft version will follow something different to what is then implemented in Keycloak.

Now, rant over and looking at what is being proposed in this PR ;)

For client authentication: Default is now changing to enforce a single audience (isser url) by default, but can be changed to be `lenient` either through two SPI options; or through configuring the authenticators.

For brokering: Default is now changed to Keycloak using issuer URL (if configured, which it in most/all cases is) instead of token URL. There doesn't seem to be any mechanism to use the old behaviour.

A few comments from me:

* Lenient is a strange name for the option; should rather be something like `allow-multiple-audiences` to make it explicit what it's for
* I don't get the need for two SPI options; why not just have a single?
* Ability to configure this through the authenticator effectively means for the whole realm; I would imagine that folks rolling this out may just be able to do it for most clients, but not for one or two that can't be easily updated. Should this not rather be a config option on individual clients?

I wonder if using a client policy to enforce this would be better; have a client policy that can enforce single audience, or allow the backwards compatible; then have a SPI option that can enable/disable this policy by default.

For identity brokering this is a breaking change now without any fallback; it basically means that this patch could not be rolled-out until all brokered IdPs are changed from expecting token url to expecting issuer url in the audience field; that is a no-go IMO.

What would be the long term plans here; when would we plan to remove the backwards compatibility options around this?

mposolda commented:
@thomasdarimont Just pointing that failure in `Base IT 1` tests is not the regression of this PR. See https://github.com/keycloak/keycloak/issues/38812

thomasdarimont commented:
Thanks Stian for your comments.

> Finally, the changes in the spec are still in a draft state; and the current [non-draft](https://openid.net/specs/openid-connect-core-1_0.html) does not include this change, which means anyone referring to the non-draft version will follow something different to what is then implemented in Keycloak.

You are right regarding, the OIDF spec, however the FAPI 2.0 Security Profil (which was just finalized in Feburary) already mentions this changed audience handling for private_key_jwt: 
See: https://openid.net/specs/fapi-security-profile-2_0-final.html#section-5.3.3.1-2.5


> For client authentication: Default is now changing to enforce a single audience (isser url) by default, but can be changed to be lenient either through two SPI options; or through configuring the authenticators.

Note that the two SPI options configure the default behavior.

> For brokering: Default is now changed to Keycloak using issuer URL (if configured, which it in most/all cases is) instead of token URL. There doesn't seem to be any mechanism to use the old behaviour.

The old behavior can only be restored by explicitly setting the `Client assertion audience` configuration config property for the OIDC Identity provider configuration.

> A few comments from me:
> Lenient is a strange name for the option; should rather be something like allow-multiple-audiences to make it explicit what it's for
> I don't get the need for two SPI options; why not just have a single?
> Ability to configure this through the authenticator effectively means for the whole realm; I would imagine that folks rolling this out may just be able to do it for most clients, but not for one or two that can't be easily updated. Should this not rather be a config option on individual clients?

Regarding the "lenient-..." name, I'm open to suggestions, but I think `allow-multiple-audiences` doesn't express the old behavior, which was to accept issuerUrl, tokenUrl, introspectionUrl, ciba authorizationCallbackUrl, etc. 
But if you all think that `allow-multiple-audiences` is better, than `lenient-audience-validation=false` I'm happy to change that.

I used two distinct configuration options to make it possible to configure it individually, since there might be scenarios where a client with private_key_jwt can support the new audience validation, but another client with client_secret_jwt cannot.

Note that the change for the JWT client authenticators is about limiting what `aud` value is accepted, and the related OIDC Identity Provider is about what `aud` value we emit in a JWT Client assertion.

> I wonder if using a client policy to enforce this would be better; have a client policy that can enforce single audience, or allow the backwards compatible; then have a SPI option that can enable/disable this policy by default.

I already discussed this with @tnorimat  and he mentioned that he might add a Client Policy which checks that the "strict" `aud` validation is used for JWT based client authentication mechanisms, as it is now also part of FAPI 2.0 SP final.

> For identity brokering this is a breaking change now without any fallback; it basically means that this patch could not be rolled-out until all brokered IdPs are changed from expecting token url to expecting issuer url in the audience field; that is a no-go IMO.

There is a fallback -> manually configuring the `Client assertion audience`as shown above. If this is not enough, we could also add a SPI config property for OIDC Identity providers similar to the JWT client authenticators above.

What would be the long term plans here; when would we plan to remove the backwards compatibility options around this?

We probably need to support the fallback for a while because clients usually adapt slowly. I guess the big identity providers / vendors will eventually catch up and support the old and new `aud` contents.

thomasdarimont commented:
I think a conservative way to move forward would be the following:
- Add support for the new stricter audience valdiation for JWT Client assertions, but don't activate it by default. However, Users can use the SPI property to enable the stricter audience validation by default.
- Announce that we will change the default to enforce the stricter audience valdiation for JWT Client assertions
- In a later Keycloak release, enforce the stricter audience validation (and deprecate the compatibility) options
- In a future Keycloak release, remove the deprecated options.

stianst commented:
Just had a conversation around this one with Ricardo and Marek. We plan to resolve the security issues around this in 26.2, but in a less intrusive way than in this PR.

In summary the draft changes to the OIDC includes 3 changes, not just one:

1. The JWT can only contain a single audience
2. The audience should be the issuer URL, and not the token URL as previously
3. When a single audience is used it must be a string, and not an array with a single element

In terms of the actual security issue only #1 is problematic. This is where if a JWT includes multiple audiences and one is malicious it can abuse the JWT to impersonate the client on the other audience.

From a security perspective #2 and #3 has no impact. #2 makes sense and you can see that from the fact that we allow not only token endpoint, but other endpoints as well. #3 on the other hand just makes no sense; there's just no reason to enforce a single string rather than an array with single entry. 

Now #2 and #3 are where there are more chances of breaking existing clients/integrations. So we will simplify the fix for 26.2 to focus only on #1 (with an SPI option to disable since there is a small risk that some may not be able to update all clients quickly).

#2 and #3 we will leave for later to deal with. #2 might be something we do as a breaking change in 27.0, while #3 may be something we just ignore, as it's a breaking change for no reason.

mposolda commented:
@thomasdarimont @stianst Created https://github.com/keycloak/keycloak/issues/38819 . I will send PR for this ASAP, so we have it ready for Keycloak 26.2 release, which is going to be released soon.

@thomasdarimont We can have this issue kept open for the follow-up work after Keycloak 26.2 (also once the OIDC specs with the strict issuer audience will be promoted to be the official OIDC specs).

mposolda commented:
Closing the PR as it was addressed differently. Keeping open the issue https://github.com/keycloak/keycloak/issues/38751 for the further alignment once the draft 3, which enforces `issuer` as an audience, is promoted to be official version of the OIDC specification.
